File: plugins/autoresearch/hooks/experiment-budget-guard.js
#!/usr/bin/env node
/**
 * PreToolUse Hook: Experiment Budget Guard
 *
 * Blocks training commands that specify a time budget exceeding the configured
 * maximum. Prevents accidental long-running experiments.
 *
 * Exit codes:
 *   0 - Allow (not a training command, or within budget)
 *   2 - Block (training command exceeds time budget)
 */

const fs = require('fs');
const path = require('path');
const MAX_STDIN = 1024 * 1024;
const DEFAULT_BUDGET = 300;

function getConfiguredBudget() {
  try {
    const configPath = path.join(process.cwd(), '.autoresearch', 'config.yaml');
    const content = fs.readFileSync(configPath, 'utf8');
    const match = content.match(/time_budget:\s*(\d+)/);
    if (match) return parseInt(match[1], 10);
  } catch {}
  return DEFAULT_BUDGET;
}

let data = '';
process.stdin.setEncoding('utf8');

process.stdin.on('data', chunk => {
  if (data.length < MAX_STDIN) data += chunk.substring(0, MAX_STDIN - data.length);
});

process.stdin.on('end', () => {
  try {
    const disabled = (process.env.DISABLED_HOOKS || '').split(',');
    if (disabled.includes('experiment-budget-guard')) {
      process.stdout.write(data);
      return;
    }

    const input = JSON.parse(data);
    const command = input.tool_input?.command || '';

    const isTraining = /\btrain\.py\b/.test(command) ||
                       /\buv\s+run\s+train\.py\b/.test(command) ||
                       /\bpython\s+.*train\.py\b/.test(command);

    if (!isTraining) {
      process.stdout.write(data);
      return;
    }

    const timeoutMatch = command.match(/--timeout\s+(\d+)/);
    if (timeoutMatch) {
      const requestedTimeout = parseInt(timeoutMatch[1], 10);
      const budget = getConfiguredBudget();
      const maxAllowed = budget * 2;

      if (requestedTimeout > maxAllowed) {
        console.error(`[Hook] BLOCKED: Training timeout ${requestedTimeout}s exceeds max allowed ${maxAllowed}s (budget: ${budget}s).`);
        console.error(`[Hook] Adjust .autoresearch/config.yaml time_budget if you need longer runs.`);
        process.stdout.write(data);
        process.exit(2);
      }
    }

    process.stdout.write(data);
  } catch (err) {
    console.error(`[Hook] experiment-budget-guard error: ${err.message}`);
    process.stdout.write(data);
  }
});

File: plugins/autoresearch/hooks/experiment-budget-guard.py
#!/usr/bin/env python3
"""
PreToolUse Hook: Experiment Budget Guard

Blocks training commands that specify a time budget exceeding the configured
maximum. Prevents accidental long-running experiments.

Exit codes:
  0 - Allow (not a training command, or within budget)
  2 - Block (training command exceeds time budget)
"""

import json
import os
import re
import sys

MAX_STDIN = 1024 * 1024
DEFAULT_BUDGET = 300


def get_configured_budget():
    try:
        config_path = os.path.join(os.getcwd(), ".autoresearch", "config.yaml")
        with open(config_path, "r") as f:
            content = f.read()
        match = re.search(r"time_budget:\s*(\d+)", content)
        if match:
            return int(match.group(1))
    except (FileNotFoundError, IOError):
        pass
    return DEFAULT_BUDGET


def main():
    data = sys.stdin.read(MAX_STDIN)

    try:
        disabled = os.environ.get("DISABLED_HOOKS", "").split(",")
        if "experiment-budget-guard" in disabled:
            sys.stdout.write(data)
            return

        inp = json.loads(data)
        command = inp.get("tool_input", {}).get("command", "")

        is_training = bool(
            re.search(r"\btrain\.py\b", command)
            or re.search(r"\buv\s+run\s+train\.py\b", command)
            or re.search(r"\bpython\s+.*train\.py\b", command)
        )

        if not is_training:
            sys.stdout.write(data)
            return

        timeout_match = re.search(r"--timeout\s+(\d+)", command)
        if timeout_match:
            requested_timeout = int(timeout_match.group(1))
            budget = get_configured_budget()
            max_allowed = budget * 2

            if requested_timeout > max_allowed:
                print(f"[Hook] BLOCKED: Training timeout {requested_timeout}s exceeds max allowed {max_allowed}s (budget: {budget}s).", file=sys.stderr)
                print("[Hook] Adjust .autoresearch/config.yaml time_budget if you need longer runs.", file=sys.stderr)
                sys.stdout.write(data)
                sys.exit(2)

        sys.stdout.write(data)
    except Exception as err:
        print(f"[Hook] experiment-budget-guard error: {err}", file=sys.stderr)
        sys.stdout.write(data)


if __name__ == "__main__":
    main()
